chore(buffering): Handle buffering messages sequentially and add more docs

olksdr opened the pull request:
This is just small fixes and mostly adding more docs about how sqlite behaves. 

This also should makes sure we handle all the incoming messages to the buffer service sequentially, which makes sure that we always have `write` before `read`.  


related to https://github.com/getsentry/team-ingest/issues/53

#skip-changelog

olksdr committed:
chore(buffering): Improve docs, and fix missed counter

jan-auer commented on relay-server/src/actors/project_buffer.rs:
```diff
@@ -296,6 +296,16 @@ impl BufferService {
+                    // Once we started spooling our in-memory buffer, the dequeue from the disk
+                    // will be blocked till this operations either fails, or commits.
+                    if let Err(err) = sqlx::query("BEGIN").execute(&db).await {
+                        relay_log::error!("failed to initiate a transaction: {}", LogError(&err));
```
Shouldn't you exit early if begin fails? I'd refactor this to an async function that returns an error and then catch it in the spawn to log errors.

jan-auer commented on relay-server/src/actors/project_buffer.rs:
```diff
@@ -296,6 +296,16 @@ impl BufferService {
                 tokio::spawn(async move {
+                    // A RESERVED lock will be acquired when the first INSERT, UPDATE, or DELETE statement is executed.
+                    // Once we started spooling our in-memory buffer, the dequeue from the disk
+                    // will be blocked till this operations either fails, or commits.
```
Where does it do that blocking?

olksdr replied:
Another query, which will try to access this table will be blocked till the lock is released. This is also partially address the `fence` suggestion from yesterday. There will be definitely only 1 query at a time.

olksdr committed:
Small refactor

olksdr committed:
Call begin transaction and commit outside of the func

olksdr committed:
Make sure the inserts done inside of the same connection (inside of 1 transaction)

iker-barriocanal commented on relay-server/src/actors/project_buffer.rs:
```diff
@@ -290,55 +332,39 @@ impl BufferService {
-                });
+
+                        // The SQL command "COMMIT" does not actually commit the changes to disk. It just turns autocommit back on.
+                        // Then, at the conclusion of the command, the regular autocommit logic takes over and causes the actual commit to disk to occur.
```
I'm not sure if I understand this correctly; but if I do, this comment is not correct. Since we're using the [WAL journal mode](https://www.sqlite.org/wal.html#how_wal_works), a `COMMIT` will [add a checkpoint](https://www.sqlite.org/wal.html#automatic_checkpoint) (so committing the changes to disk?). Then, indeed, the default committing policy is again in-place.

olksdr replied:
I do not really understand your comment here. 
The comment for this is taken from the docs, just to make sure to clarify what's happening while we call `commit` on the open transaction.

iker-barriocanal replied:
Do you mind linking to these docs then? Because I'm misunderstanding something.

olksdr replied:
https://www.sqlite.org/lockingv3.html, see transaction control.

iker-barriocanal replied:
Thanks! I misunderstood what checkpointing was, thus my confusion. It makes sense.

iker-barriocanal reviewed (commented):
Other than the comment below, the rest looks good to me.

olksdr committed:
Merge branch 'master' into chore/few-small-followups

olksdr committed:
Make sure to have read before write

Removing `tokio::spawn` from the `try_spool` function will guarantee
that spool will always happen before the read.
All the incoming messages to ProjectBuffer are processed sequentially.

olksdr committed:
Add more comments and downgraded sync from FULL to NORMAL

jjbayer commented on relay-server/src/actors/project_buffer.rs:
```diff
@@ -263,10 +270,51 @@ impl BufferService {
+        });
+
+        // Since we have 3 variables we have to bind, we devide the SQLite limit by 3
+        // here to prepare the chnunks which will be preparing the batch inserts.
```
```suggestion
        // here to prepare the chunks which will be preparing the batch inserts.
```

jjbayer commented on relay-server/src/actors/project_buffer.rs:
```diff
@@ -263,10 +270,51 @@ impl BufferService {
+
     /// Tries to save in-memory buffer to disk.
     ///
     /// It will spool to disk only if the persistent storage enabled in the configuration.
```
```suggestion
    /// It will spool to disk only if the persistent storage enabled in the configuration
    /// and a db connection can be acquired.
```

jjbayer commented on relay-server/src/actors/project_buffer.rs:
```diff
@@ -290,55 +338,26 @@ impl BufferService {
+                    // The SQL command "COMMIT" does not actually commit the changes to disk. It just turns autocommit back on.
+                    // Then, at the conclusion of the command, the regular autocommit logic takes over and causes the actual commit to disk to occur.
+                    // TODO: re-try failed commit?
+                    sqlx::query("COMMIT").execute(&mut connection).await?;
```
It looks like `sqlx` has a RAII wrapper for this: https://docs.rs/sqlx/latest/sqlx/struct.Transaction.html

Not sure if you want the automatic rollback in failure cases though.

olksdr replied:
This explicit open/close transaction has to go away now. And we will rely on the pool to do the right thing, since we handle all the messages sequentially now

jjbayer commented on relay-server/src/actors/project_buffer.rs:
```diff
@@ -380,6 +399,13 @@ impl BufferService {
         loop {
+            // By default this creates a prepared statement which is cached and re-used.
+            //
+            // Removing envelopes from the on-disk buffer in batches has following implicatations:
```
```suggestion
            // Removing envelopes from the on-disk buffer in batches has following implications:
```

iker-barriocanal replied:
(also adding "the":)
```suggestion
            // Removing envelopes from the on-disk buffer in batches has the following implications:
```

jjbayer commented on relay-server/src/actors/project_buffer.rs:
```diff
@@ -290,55 +338,26 @@ impl BufferService {
+                    // Then, at the conclusion of the command, the regular autocommit logic takes over and causes the actual commit to disk to occur.
+                    // TODO: re-try failed commit?
+                    sqlx::query("COMMIT").execute(&mut connection).await?;
+                }
```
Is it ok to do nothing if the connection cannot be acquired?

iker-barriocanal replied:
IMO `try_spool` should return an error in that case. If I call `try_spool` and it doesn't return an error, I'd expect I spooled. If there's any inner detail (i.e. connection could not be acquired), I'd expect I get the error so that I can choose to retry again if I want to.

jan-auer commented on relay-server/src/actors/project_buffer.rs:
```diff
@@ -290,55 +338,26 @@ impl BufferService {
                     return Err(BufferError::Full(estimated_db_size));
```
As a follow-up, we should: 
- try to spool as much as possible to fill up the disk (spool as many envelopes as possible)
- potentially: keep the rest in memory, and let the upstream service handle the case that we can no longer buffer. this must not "block" the caller on pending spooling, though.
- propagate an error up, via healthcheck and as 503 on queueing in the common endpoint handler.

olksdr replied:
As agreed, adding this to followups and will jump one some of the items after merging this PR.

jan-auer commented on relay-server/src/actors/project_buffer.rs:
```diff
@@ -263,10 +270,51 @@ impl BufferService {
+        // Since we have 3 variables we have to bind, we devide the SQLite limit by 3
+        // here to prepare the chnunks which will be preparing the batch inserts.
+        let mut envelopes = stream::iter(envelopes).chunks(SQLITE_LIMIT_VARIABLE_NUMBER / 3);
+        while let Some(chunk) = envelopes.next().await {
```
From offline conversation: Intention here was to not use itertools. However, using some async function on a synchronous iterator from the `futures` crate doesn't buy us anything. Let's use itertools here.

jan-auer commented on relay-server/src/actors/project_buffer.rs:
```diff
@@ -263,10 +270,51 @@ impl BufferService {
+        // Flatten all the envelopes
+        let envelopes = buffer.into_iter().flat_map(|(k, vals)| {
+            vals.into_iter()
+                .map(move |v| (k, v.into_envelope().to_vec()))
```
This loop can take quite a while, and we're not yielding. This means that the thread on the main runtime that runs this future is going to be blocked for a long time. Should we either yield here, or defer this into a `spawn_blocking`? The latter is semantically wrong, and the CPU time has to be spent, but it should be spent in a context where we do not block other futures for too long.

Example:

```rust
let mut envelopes = Vec::new();
for k, vals in buffer {
  for val in vals {
    envelopes.insert((k, val.into_envelope().to_vec()));
    tokio::task::yield_now().await;  // TODO: consider doing this every Nth iteration
  }
}
```

jan-auer commented on relay-server/src/actors/project_buffer.rs:
```diff
@@ -380,6 +399,13 @@ impl BufferService {
+            // 2. Make sure that if we panic and deleted envelopes cannot be read out fully, we do not lose all of them,
+            // but only one batch, and the rest of them will stay on disk for the next iteration
+            // to pick up.
             let mut envelopes = sqlx::query(
```
From offline conversation: Provided this could be a DELTE without subquery if a specific feature is enabled, let's add that dependency ourselves, enable the feature, and leave a comment in cargo toml, why we have it.

olksdr replied:
This is the related issue https://github.com/rusqlite/rusqlite/issues/1111, which explains the problem with this. 
If we want to enable this feature, we will have to build `sqlite` on our own. There is no easy way to just enable the feature.

jan-auer commented on relay-server/src/actors/project_buffer.rs:
```diff
@@ -519,14 +546,22 @@ impl Drop for BufferService {
+                        relay_log::error!("failed to get connection while dropping the project buffer");
+                        return;
+                    };
+                    if let Err(err) = Self::do_spool(&mut connection, buffer).await {
```
We drop this service mostly when the runtime stops, which means we should no longer start async tasks. Also, our expectation would be that when we drop the service, we close the DB and it's connections, which now isn't true because the DB lives on in a spawned task. 

This kind of procedure should probably be implemented in response to the `shutdown` signal.

For the MVP, it's enough if we drop the buffer and log outcomes + errors.

jan-auer reviewed (approved):
Approved provided that open comments are addressed. Thanks!

olksdr committed:
Merge branch 'master' into chore/few-small-followups

iker-barriocanal commented on relay-server/src/actors/project_buffer.rs:
```diff
@@ -232,6 +234,10 @@ impl BufferService {
                 // 3. Disk I/O operations tends to be more sequential using WAL.
                 // 4. WAL uses many fewer fsync() operations and is thus less vulnerable to problems on systems where the fsync() system call is broken.
                 .journal_mode(SqliteJournalMode::Wal)
+                // WAL mode is safe from corruption with synchronous=NORMAL.
+                // When synchronous is NORMAL, the SQLite database engine will still sync at the most critical moments, but less often than in FULL mode.
+                // Which guarantees good balance between safety and speed.
+                .synchronous(SqliteSynchronous::Normal)
```
I also find the following relevant, from the [docs](https://www.sqlite.org/pragma.html#:~:text=A%20transaction%20committed%20in%20WAL%20mode%20with%20synchronous%3DNORMAL%20might%20roll%20back%20following%20a%20power%20loss%20or%20system%20crash.):

> A transaction committed in WAL mode with synchronous=NORMAL might roll back following a power loss or system crash.

iker-barriocanal commented on relay-server/src/actors/project_buffer.rs:
```diff
@@ -407,6 +400,12 @@ impl BufferService {
         loop {
             // By default this creates a prepared statement which is cached and re-used.
+            //
+            // Removing envelopes from the on-disk buffer in batches has following implicatations:
```
nit: I'm not sure if I like the "on-disk" here since this in the end is an abstraction, but let's move forward with this for now.

olksdr committed:
address the pr review
